crypto/pkcs11key/key.go

```diff
@@ -223,39 +282,133 @@ func (ps *Key) getPrivateKey(module ctx, session pkcs11.SessionHandle, label str
 }
 
 // Get the public key matching a private key
-// TODO: Add support for non-RSA keys, switching on CKA_KEY_TYPE
-func getPublicKey(module ctx, session pkcs11.SessionHandle, privateKeyHandle pkcs11.ObjectHandle) (rsa.PublicKey, error) {
-	var noKey rsa.PublicKey
-	template := []*pkcs11.Attribute{
-		pkcs11.NewAttribute(pkcs11.CKA_MODULUS, nil),
-		pkcs11.NewAttribute(pkcs11.CKA_PUBLIC_EXPONENT, nil),
-	}
-	attr, err := module.GetAttributeValue(session, privateKeyHandle, template)
-	if err != nil {
-		return noKey, err
-	}
-
-	n := big.NewInt(0)
-	e := int(0)
-	gotModulus, gotExponent := false, false
-	for _, a := range attr {
-		if a.Type == pkcs11.CKA_MODULUS {
-			n.SetBytes(a.Value)
-			gotModulus = true
-		} else if a.Type == pkcs11.CKA_PUBLIC_EXPONENT {
-			bigE := big.NewInt(0)
-			bigE.SetBytes(a.Value)
-			e = int(bigE.Int64())
-			gotExponent = true
+func (ps *Key) getPublicKey(module ctx, session pkcs11.SessionHandle, privateKeyHandle pkcs11.ObjectHandle) (interface{}, error) {
```

Comment:
This is now a huge function. It might be better to split it into getRSAPublicKey and getECPublicKey.